File: lab6/lab6.m
close all;
clear;

IMG = imread('4demosaicking.bmp');
out = demosaic(IMG);
image(out)

%imwrite(v_nearest(IMG, vector, 0), 'v_nearest_1.png'


function out = demosaic(in)
    Xo = length(in(1,:,1));
    Yo = length(in(:,1,1));
    
    for y = 1:Yo
        for x = 1:Xo
            if mod(Xo,2) == 1
                if mod(Yo,2) == 1
                    out(y,x,1) = in(y,x,1);
                    out(y,x,2) = uint8(1);
                    out(y,x,3) = uint8(1);
                else
                    out(y,x,1) = uint8(1)
                    out(y,x,2) = in(y,x,2);
                    out(y,x,3) = uint8(1);
                end
            else
                if mod(Yo,2) == 1
                    out(y,x,1) = uint8(1)
                    out(y,x,2) = in(y,x,2);
                    out(y,x,3) = uint8(1);
                else
                    out(y,x,1) = uint8(1);
                    out(y,x,2) = uint8(1);
                    out(y,x,3) = in(y,x,3);
                end
            end
        end
    end
end

File: lab6/lab6.py
#!/usr/bin/env python

import matplotlib.image as mpimg 
import matplotlib.pyplot as plt
import numpy as np

def mosaic(img, mask):
    mask_lx = len(mask[0])
    mask_ly = len(mask)
    img_lx = len(img[0])
    img_ly = len(img)
    out = np.array([[0]*len(img[0])]*len(img), dtype='float')
    #out = [[0 for i in range(10)] for j in range(20)] 
    for y in range(img_ly):
        for x in range(img_lx):
            out[y][x] = (img[y][x][0]*mask[y%mask_ly][x%mask_lx][0] +
                         img[y][x][1]*mask[y%mask_ly][x%mask_lx][1] +
                         img[y][x][2]*mask[y%mask_ly][x%mask_lx][2])
    
    return out


def BAYER_MASK():
    return [
        [[0,1,0], [1,0,0]],
        [[0,0,1], [0,1,0]]
            ]

def XTRANS_MASK():
    return [
        [[0,1,0], [0,0,1], [0,1,0],   [0,1,0], [1,0,0], [0,1,0]],
        [[1,0,0], [0,1,0], [1,0,0],   [0,0,1], [0,1,0], [0,0,1]],
        [[0,1,0], [0,0,1], [0,1,0],   [0,1,0], [1,0,0], [0,1,0]],
        
        [[0,1,0], [1,0,0], [0,1,0],   [0,1,0], [0,0,1], [0,1,0]],
        [[0,0,1], [0,1,0], [0,0,1],   [1,0,0], [0,1,0], [1,0,0]],
        [[0,1,0], [1,0,0], [0,1,0],   [0,1,0], [0,0,1], [0,1,0]],
            ]


img = mpimg.imread('4demosaicking.png') 

plt.imshow(mosaic(img, XTRANS_MASK()), cmap='gray',  interpolation='nearest') 
plt.show()


#print (img.shape)
#plt.imshow(img) 
#plt.show()
